From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Wed, 6 Jan 2021 14:11:02 -0500
Subject: [PATCH] Disable the loading of certain plugins


diff --git a/src/main/java/me/titaniumtown/TitaniumUtils.java b/src/main/java/me/titaniumtown/TitaniumUtils.java
new file mode 100644
index 0000000000000000000000000000000000000000..5187b3b23bc63bfcf1c718f4987550f8c5fa3309
--- /dev/null
+++ b/src/main/java/me/titaniumtown/TitaniumUtils.java
@@ -0,0 +1,12 @@
+package me.titaniumtown;
+
+import com.google.common.collect.ImmutableSet;
+
+public final class TitaniumUtils {
+  public static final ImmutableSet<String> BANNED_MAIN_CLASSES = ImmutableSet.of(
+      "net.minemora.entitytrackerfixer.EntityTrackerFixer", // EntityTrackerFixer
+      "me.minebuilders.clearlag.Clearlag", // Clearlag
+      "com.froobworld.avl.Avl", // Villager Optimizer
+      "com.songoda.ultimatetimber.UltimateTimber"
+  );
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index 26685f59b235ea5b4c4fb7ae21acb5149edaa2b3..9739d6e740f897a850979c80b41a1dac22bb2c05 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -147,6 +147,10 @@ public final class SimplePluginManager implements PluginManager {
             try {
                 description = loader.getPluginDescription(file);
                 String name = description.getName();
+                if (name.equalsIgnoreCase("FastAsyncWorldEdit") && !Boolean.getBoolean("Titanium.allowAllPlugins")) {
+                    server.getLogger().log(Level.WARNING, "Skipping loading of FastAsyncWorldEdit! Please use regular WorldEdit (or AWE) instead as FAWE is riddled with bugs");
+                    continue;
+                }
                 if (name.equalsIgnoreCase("bukkit") || name.equalsIgnoreCase("minecraft") || name.equalsIgnoreCase("mojang")) {
                     server.getLogger().log(Level.SEVERE, "Could not load '" + file.getPath() + "' in folder '" + directory.getPath() + "': Restricted Name");
                     continue;
diff --git a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
index 04fa3991f6ce4e9dad804f28fc6c947695857089..ae836980d3e82ef6fe87d641fcf6766eb759a293 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
@@ -1,6 +1,7 @@
 package org.bukkit.plugin.java;
 
 import com.google.common.base.Charsets;
+import me.titaniumtown.TitaniumUtils;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
@@ -257,6 +258,12 @@ public abstract class JavaPlugin extends PluginBase {
      */
     protected final void setEnabled(final boolean enabled) {
         if (isEnabled != enabled) {
+            // Titanium start
+            if (TitaniumUtils.BANNED_MAIN_CLASSES.contains(getDescription().getMain()) && !Boolean.getBoolean("Titanium.allowAllPlugins")) {
+                getLogger().warning(getDescription().getName() + " has been disabled as it's not recommended to use.");
+                return;
+            }
+            // Titanium end
             isEnabled = enabled;
 
             if (isEnabled) {
