From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Wed, 2 Dec 2020 14:56:35 -0500
Subject: [PATCH] remove snooper


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 9399b3c1571dfbbc58476e3d92599f589fa0094a..585a02f1bae2cfb13c7efc7c5e2b2eb95fdb8344 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1317,14 +1317,6 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         //} // Paper
 
         this.methodProfiler.enter("snooper");
-        if (((DedicatedServer) this).getDedicatedServerProperties().snooperEnabled && !this.snooper.d() && this.ticks > 100) { // Spigot
-            this.snooper.a();
-        }
-
-        if (((DedicatedServer) this).getDedicatedServerProperties().snooperEnabled && this.ticks % 6000 == 0) { // Spigot
-            this.snooper.b();
-        }
-
         this.methodProfiler.exit();
 
         // Paper start - move executeAll() into full server tick timing
