From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Tue, 15 Dec 2020 20:42:22 -0500
Subject: [PATCH] allow toggle for sand dupe


diff --git a/src/main/java/me/titaniumtown/misc.java b/src/main/java/me/titaniumtown/misc.java
index e8048e0bb..aad757414 100644
--- a/src/main/java/me/titaniumtown/misc.java
+++ b/src/main/java/me/titaniumtown/misc.java
@@ -19,3 +19,4 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
+public static boolean allowSandDupe = true;
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/EntityFallingBlock.java b/src/main/java/net/minecraft/server/EntityFallingBlock.java
index 76aac1b13..a90e76e80 100644
--- a/src/main/java/net/minecraft/server/EntityFallingBlock.java
+++ b/src/main/java/net/minecraft/server/EntityFallingBlock.java
@@ -64,8 +64,8 @@ public class EntityFallingBlock extends Entity {
 
     @Override
     public void tick() {
-        // Paper start - fix sand duping
-        if (this.dead) {
+        // Paper start - fix sand duping // Titanium - add config option for sand duping
+        if (this.dead && !me.titaniumtown.misc.allowSandDupe) {
             return;
         }
         // Paper end - fix sand duping
@@ -91,8 +91,8 @@ public class EntityFallingBlock extends Entity {
 
             this.move(EnumMoveType.SELF, this.getMot());
 
-            // Paper start - fix sand duping
-            if (this.dead) {
+            // Paper start - fix sand duping // Titanium - add config option for sand duping
+            if (this.dead && !me.titaniumtown.misc.allowSandDupe) {
                 return;
             }
             // Paper end - fix sand duping
