From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Wed, 16 Dec 2020 10:59:57 -0500
Subject: [PATCH] add config for toggling sand dupe fix


diff --git a/src/main/java/me/titaniumtown/TitaniumConfig.java b/src/main/java/me/titaniumtown/TitaniumConfig.java
index 9bb47ab46b1ae95888e395bfea50cd6afee5bb8c..663f033bd2ed3fe4bd18dc562f0a26a01b6fe6de 100644
--- a/src/main/java/me/titaniumtown/TitaniumConfig.java
+++ b/src/main/java/me/titaniumtown/TitaniumConfig.java
@@ -153,4 +153,9 @@ public class TitaniumConfig {
         return config.getString(path, config.getString(path));
     }
 
+    public static boolean allowSandDupe = false;
+    private static void experimental() {
+        allowSandDupe = getBoolean("settings.experimental.allow-sand-dupe", allowSandDupe);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/EntityFallingBlock.java b/src/main/java/net/minecraft/server/EntityFallingBlock.java
index 76aac1b131f314775e418339e434f4f2da2ad619..7236d5112c2d4d6749b837ea8dbfcd89dfe9dd7d 100644
--- a/src/main/java/net/minecraft/server/EntityFallingBlock.java
+++ b/src/main/java/net/minecraft/server/EntityFallingBlock.java
@@ -64,8 +64,8 @@ public class EntityFallingBlock extends Entity {
 
     @Override
     public void tick() {
-        // Paper start - fix sand duping
-        if (this.dead) {
+        // Paper start - fix sand duping // Titanium - add config option for sand duping
+        if (this.dead && !me.titaniumtown.TitaniumConfig.allowSandDupe) {
             return;
         }
         // Paper end - fix sand duping
@@ -91,8 +91,8 @@ public class EntityFallingBlock extends Entity {
 
             this.move(EnumMoveType.SELF, this.getMot());
 
-            // Paper start - fix sand duping
-            if (this.dead) {
+            // Paper start - fix sand duping // Titanium - add config option for sand duping
+            if (this.dead && !me.titaniumtown.TitaniumConfig.allowSandDupe) {
                 return;
             }
             // Paper end - fix sand duping
