From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Wed, 16 Dec 2020 14:50:38 -0500
Subject: [PATCH] add config for farlands


diff --git a/src/main/java/me/titaniumtown/TitaniumConfig.java b/src/main/java/me/titaniumtown/TitaniumConfig.java
index 178d588fe..acd41c672 100644
--- a/src/main/java/me/titaniumtown/TitaniumConfig.java
+++ b/src/main/java/me/titaniumtown/TitaniumConfig.java
@@ -160,10 +160,11 @@ public class TitaniumConfig {
 
     public static double WborderSize = 6.0E7D; // Vanilla 6.0E7D; Max: 4.294967294E9D
     public static int biggestInt = 29999984; // Vanilla: 29999984; Max: 2147483647
+    public static boolean farlands = false;
     private static void world() {
         WborderSize = getDouble("settings.experimental.world.world-border-size", WborderSize);
         biggestInt = getInt("settings.experimental.world.world-max-int", biggestInt);
-
+        farlands = getBoolean("settings.experimental.world.enable-farlands", farlands);
         // Start - World size settings checks
         if (WborderSize > 4.294967294E9D) Bukkit.getLogger().log(Level.SEVERE, "Titanium: settings.experimental.world.world-border-size is set over it's maximum value");
         if (biggestInt > 2147483647) Bukkit.getLogger().log(Level.SEVERE, "Titanium: settings.experimental.world.world-max-int is set over it's maximum value");
diff --git a/src/main/java/net/minecraft/server/NoiseGeneratorOctaves.java b/src/main/java/net/minecraft/server/NoiseGeneratorOctaves.java
index 1b7454609..008207e22 100644
--- a/src/main/java/net/minecraft/server/NoiseGeneratorOctaves.java
+++ b/src/main/java/net/minecraft/server/NoiseGeneratorOctaves.java
@@ -142,7 +142,11 @@ public class NoiseGeneratorOctaves implements NoiseGenerator {
     }
 
     public static double a(double d0) {
-        return d0 - (double) MathHelper.d(d0 / 3.3554432E7D + 0.5D) * 3.3554432E7D;
+        if (me.titaniumtown.TitaniumConfig.farlands) {
+            return d0;
+        } else {
+            return d0 - (double) MathHelper.d(d0 / 3.3554432E7D + 0.5D) * 3.3554432E7D;
+        }
     }
 
     @Override
