From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Thu, 17 Dec 2020 16:06:49 -0500
Subject: [PATCH] remove some coords checks


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 9d643df31ccadf082889f3a502ed139d3b8ec569..66bc97ac4ee9d70c2a01053a823934c6b8bb8a63 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -341,7 +341,9 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
     }
 
     private static boolean D(BlockPosition blockposition) {
-        return blockposition.getX() >= -me.titaniumtown.TitaniumConfig.replace3 && blockposition.getZ() >= -me.titaniumtown.TitaniumConfig.replace3 && blockposition.getX() < me.titaniumtown.TitaniumConfig.replace3 && blockposition.getZ() < me.titaniumtown.TitaniumConfig.replace3; // Titanium - Remove world border limitations
+        // return blockposition.getX() >= -me.titaniumtown.TitaniumConfig.replace3 && blockposition.getZ() >= -me.titaniumtown.TitaniumConfig.replace3 && blockposition.getX() < me.titaniumtown.TitaniumConfig.replace3 && blockposition.getZ() < me.titaniumtown.TitaniumConfig.replace3; // Titanium - Remove world border limitations // Titanium - Remove movement checks
+        return true;
+
     }
 
     private static boolean d(int i) {
@@ -729,15 +731,13 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
         int k;
 
         // Titanium start - Remove world border limitations
-        if (i >= -me.titaniumtown.TitaniumConfig.replace3 && j >= -me.titaniumtown.TitaniumConfig.replace3 && i < me.titaniumtown.TitaniumConfig.replace3 && j < me.titaniumtown.TitaniumConfig.replace3) {
-            if (this.isChunkLoaded(i >> 4, j >> 4)) {
-                k = this.getChunkAt(i >> 4, j >> 4).getHighestBlock(heightmap_type, i & 15, j & 15) + 1;
-            } else {
-                k = 0;
-            }
+        // Titanium start - Remove movement checks
+        if (this.isChunkLoaded(i >> 4, j >> 4)) {
+            k = this.getChunkAt(i >> 4, j >> 4).getHighestBlock(heightmap_type, i & 15, j & 15) + 1;
         } else {
-            k = this.getSeaLevel() + 1;
+            k = 0;
         }
+        // Titanium start - Remove movement checks
         // Titanium end
 
         return k;
