From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sotr <i@omc.hk>
Date: Wed, 15 Apr 2020 22:45:48 +0700
Subject: [PATCH] (Akarin) Remove a few more streams


diff --git a/src/main/java/net/minecraft/server/ChunkMapDistance.java b/src/main/java/net/minecraft/server/ChunkMapDistance.java
index f1c686810fb4e9c05df45d664c93af73d17f0624..b4133720127163157058f73c2e3270f055ca87e4 100644
--- a/src/main/java/net/minecraft/server/ChunkMapDistance.java
+++ b/src/main/java/net/minecraft/server/ChunkMapDistance.java
@@ -192,9 +192,15 @@ public abstract class ChunkMapDistance {
                 while (longiterator.hasNext()) {
                     long j = longiterator.nextLong();
 
+                    // Akarin start - remove stream
+                    ArraySetSorted<Ticket<?>> sorted = this.tickets.get(j);
+                    /*
                     if (this.e(j).stream().anyMatch((ticket) -> {
                         return ticket.getTicketType() == TicketType.PLAYER;
                     })) {
+                    */
+                    Runnable runnable = () -> {
+                    // Akarin end
                         PlayerChunk playerchunk = playerchunkmap.getUpdatingChunk(j);
 
                         if (playerchunk == null) {
@@ -210,6 +216,22 @@ public abstract class ChunkMapDistance {
                             });
                         });
                     }
+                    // Akarin start - remove stream
+                    ; // end runnable
+                    
+                    if (sorted == null) {
+                      sorted = ArraySetSorted.a(4);
+                      this.tickets.put(j, sorted);
+                      
+                    } else {
+                      for (Ticket<?> ticket : sorted) {
+                        if (ticket.getTicketType() == TicketType.PLAYER) {
+                          runnable.run();
+                          break; // run for only once
+                        }
+                      }
+                    }
+                    // Akarin end
                 }
 
                 this.l.clear();
diff --git a/src/main/java/net/minecraft/server/EntityTNTPrimed.java b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
index b108874d977a4fb1a7c0d52f66bc08390decb4d0..35b0cb353ce608471c30df6c0ee854fdcdc1a26d 100644
--- a/src/main/java/net/minecraft/server/EntityTNTPrimed.java
+++ b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
@@ -92,12 +92,23 @@ public class EntityTNTPrimed extends Entity {
                 PacketPlayOutEntityVelocity velocityPacket = new PacketPlayOutEntityVelocity(this);
                 PacketPlayOutEntityTeleport positionPacket = new PacketPlayOutEntityTeleport(this);
 
+                // Akarin start - remove stream
+                /*
                 ete.trackedPlayers.stream()
                     .filter(viewer -> (viewer.locX() - this.locX()) * (viewer.locY() - this.locY()) * (viewer.locZ() - this.locZ()) < 16 * 16)
                     .forEach(viewer -> {
                         viewer.playerConnection.sendPacket(velocityPacket);
                         viewer.playerConnection.sendPacket(positionPacket);
                     });
+                */
+                
+                for (EntityPlayer viewer : ete.trackedPlayers) {
+                  if ((viewer.locX() - this.locX()) * (viewer.locY() - this.locY()) * (viewer.locZ() - this.locZ()) < 16 * 16) {
+                    viewer.playerConnection.sendPacket(velocityPacket);
+                    viewer.playerConnection.sendPacket(positionPacket);
+                  }
+                }
+                // Akarin end
             }
         }
         // Paper end
diff --git a/src/main/java/net/minecraft/server/EntityVillager.java b/src/main/java/net/minecraft/server/EntityVillager.java
index 08edab66c18200dfea29b52e2825af94119a3415..a6ed9d6e02b620bcaed4e63bd3b2c764618f2d71 100644
--- a/src/main/java/net/minecraft/server/EntityVillager.java
+++ b/src/main/java/net/minecraft/server/EntityVillager.java
@@ -650,11 +650,20 @@ public class EntityVillager extends EntityVillagerAbstract implements Reputation
             if (optional.isPresent()) {
                 WorldServer worldserver = (WorldServer) this.world;
 
+                // Akarin start - remote stream
+                /*
                 ((List) optional.get()).stream().filter((entityliving) -> {
                     return entityliving instanceof ReputationHandler;
                 }).forEach((entityliving) -> {
                     worldserver.a(ReputationEvent.d, entity, (ReputationHandler) entityliving);
                 });
+                */
+                
+                for (EntityLiving each : optional.get()) {
+                  if (each instanceof ReputationHandler)
+                    worldserver.a(ReputationEvent.d, entity, (ReputationHandler) each);
+                }
+                // Akarin end
             }
         }
     }
