From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Tue, 22 Dec 2020 00:53:58 -0500
Subject: [PATCH] add more dupe options


diff --git a/src/main/java/me/titaniumtown/TitaniumConfig.java b/src/main/java/me/titaniumtown/TitaniumConfig.java
index 5bd706ab08efde4fa0e953cdb9caae164ff042dc..3078eaa6a1925cf00bbdbb3cab9dfb53c091ca01 100644
--- a/src/main/java/me/titaniumtown/TitaniumConfig.java
+++ b/src/main/java/me/titaniumtown/TitaniumConfig.java
@@ -154,8 +154,12 @@ public class TitaniumConfig {
     }
 
     public static boolean allowSandDupe = false;
+    public static boolean allowPlayerItemDuplication = false;
+    public static boolean allowRidableChestDuping = false;
     private static void experimental() {
         allowSandDupe = getBoolean("settings.experimental.allow-sand-dupe", allowSandDupe);
+        allowPlayerItemDuplication = getBoolean("settings.experimental.allow-player-item-dupe", allowPlayerItemDuplication);
+        allowRidableChestDuping = getBoolean("settings.experimental.allow-ridable-chest-dupe", allowRidableChestDuping);
     }
 
     public static double WborderSize = 6.0E7D; // Vanilla 6.0E7D; Max: 4.294967294E9D // replacement for 6.0E7D
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 9d8facf395fe3dc427da46c7177d32fb3abc84d8..95ab987b4123e23001f3dcf212e7dd03492d3191 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -2156,6 +2156,7 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
             }
             // CraftBukkit end
             EntityItem entityitem = new EntityItem(this.world, this.locX(), this.locY() + (double) f, this.locZ(), itemstack.cloneItemStack()); // Paper - clone so we can destroy original
+            if (!me.titaniumtown.TitaniumConfig.allowRidableChestDuping) // Titanium
             itemstack.setCount(0); // Paper - destroy this item - if this ever leaks due to game bugs, ensure it doesn't dupe
 
             entityitem.defaultPickupDelay();
@@ -2826,7 +2827,7 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
     public Entity teleportTo(WorldServer worldserver, BlockPosition location) {
         // CraftBukkit end
         // Paper start - fix bad state entities causing dupes
-        if (!isAlive() || !valid) {
+        if (!me.titaniumtown.TitaniumConfig.allowRidableChestDuping && (!isAlive() || !valid)) { // Titanium
             LOGGER.warn("Illegal Entity Teleport " + this + " to " + worldserver + ":" + location, new Throwable());
             return null;
         }
@@ -2866,7 +2867,7 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
                     entity.bukkitEntity = this.getBukkitEntity();
 
                     if (this instanceof EntityInsentient) {
-                        ((EntityInsentient) this).unleash(true, true); // Paper drop lead
+                        ((EntityInsentient) this).unleash(true, !me.titaniumtown.TitaniumConfig.allowRidableChestDuping); // Paper drop lead // Rainforest
                     }
                     // CraftBukkit end
                 }
@@ -2988,7 +2989,7 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
     }
 
     public boolean canPortal() {
-        return isAlive() && valid && (world.purpurConfig.entitiesCanUsePortals || this instanceof EntityPlayer); // Paper // Purpur
+        return me.titaniumtown.TitaniumConfig.allowRidableChestDuping || isAlive() && valid && (world.purpurConfig.entitiesCanUsePortals || this instanceof EntityPlayer); // Paper // Purpur // Titaniumtown
     }
 
     public float a(Explosion explosion, IBlockAccess iblockaccess, BlockPosition blockposition, IBlockData iblockdata, Fluid fluid, float f) {
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index cee95fa977aa7df57a411b79be688e6b949b05d5..27e444c48593a15348acd899a40533b1698afedf 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -2382,7 +2382,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
     @Override
     public boolean isFrozen() { // Paper - protected > public
-        return super.isFrozen() || frozen || (this.playerConnection != null && this.playerConnection.isDisconnected()); // Paper // Purpur
+        return super.isFrozen() || (me.titaniumtown.TitaniumConfig.allowPlayerItemDuplication ? !this.playerConnection.isDisconnected() : this.playerConnection != null && this.playerConnection.isDisconnected()); // Paper // Purpur // Titanium
     }
 
     // Purpur start
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 6e9315c92ac9ba4dec9ee68de5ce0cb7a356c106..5fd375769b33ead5e2c49ae889a4229a95d5e15c 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -2999,7 +2999,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
     // Paper end
 
     public final boolean isDisconnected() {
-        return (!this.player.joining && !this.networkManager.isConnected()) || this.processedDisconnect; // Paper
+        return (!this.player.joining && !this.networkManager.isConnected()) || me.titaniumtown.TitaniumConfig.allowPlayerItemDuplication || this.processedDisconnect; // Paper // Titanium
     }
     // CraftBukkit end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 110db570c8c917e6b9eeab534e47ce9135ad09c9..098e70a2de4479b72f3e79230ea08dea85c6dacd 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -824,6 +824,7 @@ public class CraftEventFactory {
             if (stack == null || stack.getType() == Material.AIR || stack.getAmount() == 0) continue;
 
             world.dropItem(entity.getLocation(), stack); // Paper - note: dropItem already clones due to this being bukkit -> NMS
+            if (!me.titaniumtown.TitaniumConfig.allowRidableChestDuping) // Titanium
             if (stack instanceof CraftItemStack) stack.setAmount(0); // Paper - destroy this item - if this ever leaks due to game bugs, ensure it doesn't dupe, but don't nuke bukkit stacks of manually added items
         }
 
