From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Tue, 9 Feb 2021 11:10:56 -0500
Subject: [PATCH] Revert "Fix sand duping"

This reverts commit 0477ab4ccfed8662d556d925b689ade03197e807.

diff --git a/src/main/java/net/minecraft/world/entity/item/EntityFallingBlock.java b/src/main/java/net/minecraft/world/entity/item/EntityFallingBlock.java
index 62d8b53c024888aa43b8fddf8a475dfb8284a4cc..411e3915c0aa00249aacb6658ed04309665d2fb4 100644
--- a/src/main/java/net/minecraft/world/entity/item/EntityFallingBlock.java
+++ b/src/main/java/net/minecraft/world/entity/item/EntityFallingBlock.java
@@ -103,11 +103,6 @@ public class EntityFallingBlock extends Entity {
 
     @Override
     public void tick() {
-        // Paper start - fix sand duping
-        if (this.dead) {
-            return;
-        }
-        // Paper end - fix sand duping
         if (this.block.isAir()) {
             this.die();
         } else {
@@ -130,12 +125,6 @@ public class EntityFallingBlock extends Entity {
 
             this.move(EnumMoveType.SELF, this.getMot());
 
-            // Paper start - fix sand duping
-            if (this.dead) {
-                return;
-            }
-            // Paper end - fix sand duping
-
             // Paper start - Configurable EntityFallingBlock height nerf
             if (this.world.paperConfig.fallingBlockHeightNerf != 0 && this.locY() > this.world.paperConfig.fallingBlockHeightNerf) {
                 if (this.dropItem && this.world.getGameRules().getBoolean(GameRules.DO_ENTITY_DROPS)) {
