From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Mon, 8 Feb 2021 16:11:47 -0500
Subject: [PATCH] fully remove snooper


diff --git a/src/main/java/net/minecraft/server/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/DedicatedServerProperties.java
index d2520b7ee84f1e307faca2018c75f79d75ed707d..adda2a0af7814ae04d0ce84ae4d3980061aae84a 100644
--- a/src/main/java/net/minecraft/server/DedicatedServerProperties.java
+++ b/src/main/java/net/minecraft/server/DedicatedServerProperties.java
@@ -39,7 +39,7 @@ public class DedicatedServerProperties extends PropertyManager<DedicatedServerPr
     public final boolean hardcore;
     public final boolean allowNether;
     public final boolean spawnMonsters;
-    public final boolean snooperEnabled;
+    // public final boolean snooperEnabled;
     public final boolean useNativeTransport;
     public final boolean enableCommandBlock;
     public final int spawnProtection;
@@ -86,11 +86,13 @@ public class DedicatedServerProperties extends PropertyManager<DedicatedServerPr
         this.hardcore = this.getBoolean("hardcore", false);
         this.allowNether = this.getBoolean("allow-nether", true);
         this.spawnMonsters = this.getBoolean("spawn-monsters", true);
+        /*
         if (this.getBoolean("snooper-enabled", true)) {
             ;
         }
 
         this.snooperEnabled = false;
+        */
         this.useNativeTransport = this.getBoolean("use-native-transport", true);
         this.enableCommandBlock = this.getBoolean("enable-command-block", false);
         this.spawnProtection = this.getInt("spawn-protection", 16);
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 2df2f619715f38040feff48409022957092b9c16..62c80bd91656adafa03e1bb4f01734e659aa0d77 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -79,7 +79,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
     public static final WorldSettings c = new WorldSettings("Demo World", EnumGamemode.SURVIVAL, false, EnumDifficulty.NORMAL, false, new GameRules(), DataPackConfiguration.a);
     public Convertable.ConversionSession convertable;
     public final WorldNBTStorage worldNBTStorage;
-    private final MojangStatisticsGenerator snooper = new MojangStatisticsGenerator("server", this, SystemUtils.getMonotonicMillis());
+    // private final MojangStatisticsGenerator snooper = new MojangStatisticsGenerator("server", this, SystemUtils.getMonotonicMillis());
     private final List<Runnable> tickables = Lists.newArrayList();
     private final GameProfilerSwitcher m;
     private GameProfilerFiller methodProfiler;
@@ -817,9 +817,11 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
             }
         }
 
+        /*
         if (this.snooper.d()) {
             this.snooper.e();
         }
+        */
 
         this.dataPackResources.close();
 
@@ -1339,7 +1341,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
             //MinecraftServer.LOGGER.debug("Autosave finished"); // Paper
         //} // Paper
 
-        this.methodProfiler.enter("snooper");
+        // this.methodProfiler.enter("snooper");
         // Akarin - Disable Snooper
         /*
         if (((DedicatedServer) this).getDedicatedServerProperties().snooperEnabled && !this.snooper.d() && this.ticks > 100) { // Spigot
@@ -1351,7 +1353,7 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         }
         */
 
-        this.methodProfiler.exit();
+        // this.methodProfiler.exit();
 
         // Paper start - move executeAll() into full server tick timing
         try (co.aikar.timings.Timing ignored = MinecraftTimings.processTasksTimer.startTiming()) {
