From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Fri, 5 Feb 2021 14:44:16 -0500
Subject: [PATCH] remove config for Optimized Explotions (always on)


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index c46ef87f02450340e8be6ac3936f72598a7c9088..8b2e003dc5f5539fc8f6057c383a3dc800db62d6 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -179,12 +179,6 @@ public class PaperWorldConfig {
         log("End credits disabled: " + disableEndCredits);
     }
 
-    public boolean optimizeExplosions;
-    private void optimizeExplosions() {
-        optimizeExplosions = getBoolean("optimize-explosions", false);
-        log("Optimize explosions: " + optimizeExplosions);
-    }
-
     public boolean disableExplosionKnockback;
     private void disableExplosionKnockback(){
         disableExplosionKnockback = getBoolean("disable-explosion-knockback", false);
diff --git a/src/main/java/net/minecraft/server/Explosion.java b/src/main/java/net/minecraft/server/Explosion.java
index 5ca4d03215c979a4d122a32178a1859b4d9c3334..bef7db87779a41dfa42794df1e51bba649ea8c72 100644
--- a/src/main/java/net/minecraft/server/Explosion.java
+++ b/src/main/java/net/minecraft/server/Explosion.java
@@ -404,9 +404,6 @@ public class Explosion {
     }
     // Paper start - Optimize explosions
     private float getBlockDensity(Vec3D vec3d, Entity entity) {
-        if (!this.world.paperConfig.optimizeExplosions) {
-            return a(vec3d, entity);
-        }
         CacheKey key = new CacheKey(this, entity.getBoundingBox());
         Float blockDensity = this.world.explosionDensityCache.get(key);
         if (blockDensity == null) {
