From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JellySquid <jellysquid+atwork@protonmail.com>
Date: Sun, 24 Jan 2021 22:55:55 +0100
Subject: [PATCH] lithium entity


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index bf9ab7e83f6f26ffacd005b7180ea3746a5d21f4..934b17fdc0dc05a6b0ab0eb0b881854c334e1826 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -191,6 +191,10 @@ public abstract class EntityLiving extends Entity {
     public NearbyEntityListenerMulti getListener() {
         return this.tracker;
     }
+
+    private IBlockData lastStateAtFeet = null;
+
+    private long lastPos = Long.MIN_VALUE;
     // Yatopia end
     protected void initAttributes() {} // Purpur
 
@@ -268,6 +272,10 @@ public abstract class EntityLiving extends Entity {
 
     @Override
     public void entityBaseTick() {
+        // Yatopia start - port lithium
+        this.lastStateAtFeet = null;
+        this.lastPos = Long.MIN_VALUE;
+        // Yatopia end
         this.ar = this.as;
         if (this.justCreated) {
             this.getBedPosition().ifPresent(this::a);
@@ -1691,6 +1699,26 @@ public abstract class EntityLiving extends Entity {
 
     public IBlockData ds() {
         return this.world.getBlockStateIfLoaded(this.getChunkCoordinates());
+        /*
+        // Yatopia start - port lithium
+        int x = MathHelper.floor(this.locX());
+        int y = MathHelper.floor(this.locY());
+        int z = MathHelper.floor(this.locZ());
+
+        long pos = getChunkCoordinates().asLong();
+
+        if (this.lastPos == pos) {
+            return this.lastStateAtFeet;
+        }
+
+        IBlockData state = this.world.getType(getChunkCoordinates());
+
+        this.lastPos = pos;
+        this.lastStateAtFeet = state;
+
+        return state;
+        // Yatopia end
+        */
     }
 
     private boolean c(BlockPosition blockposition, IBlockData iblockdata) {
