From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Tue, 9 Feb 2021 11:17:52 -0500
Subject: [PATCH] Revert "Fix piston physics inconsistency - MC-188840"

This reverts commit c59f6d4cc7fe84c4bdbdbb5b2c1f0bf87c34304f.

diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index fc1cd9e41e4461c626ffbef68aa0d321202c6907..38eb15ebe74a55181f5c68dedec97c9577d80442 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -441,13 +441,6 @@ public class PaperConfig {
         consoleHasAllPermissions = getBoolean("settings.console-has-all-permissions", consoleHasAllPermissions);
     }
 
-    public static boolean allowPistonDuplication;
-    private static void allowPistonDuplication() {
-        config.set("settings.unsupported-settings.allow-piston-duplication-readme", "This setting controls if player should be able to use TNT duplication, but this also allows duplicating carpet, rails and potentially other items");
-        allowPistonDuplication = getBoolean("settings.unsupported-settings.allow-piston-duplication", config.getBoolean("settings.unsupported-settings.allow-tnt-duplication", false));
-        set("settings.unsupported-settings.allow-tnt-duplication", null);
-    }
-
     public static boolean allowHeadlessPistons;
     private static void allowHeadlessPistons() {
         config.set("settings.unsupported-settings.allow-headless-pistons-readme", "This setting controls if players should be able to create headless pistons.");
diff --git a/src/main/java/net/minecraft/server/BlockPiston.java b/src/main/java/net/minecraft/server/BlockPiston.java
index 7f084e7f11a829c10d113c7fb39eec0b8ad77080..337405a5055aaee7c9bab45812c716272d686a5e 100644
--- a/src/main/java/net/minecraft/server/BlockPiston.java
+++ b/src/main/java/net/minecraft/server/BlockPiston.java
@@ -371,24 +371,12 @@ public class BlockPiston extends BlockDirectional {
             }
 
             for (k = list.size() - 1; k >= 0; --k) {
-                // Paper start - fix a variety of piston desync dupes
-                boolean allowDesync = com.destroystokyo.paper.PaperConfig.allowPistonDuplication;
-                BlockPosition oldPos = blockposition3 = (BlockPosition) list.get(k);
-                iblockdata1 = allowDesync ? world.getType(oldPos) : null;
-                // Paper end - fix a variety of piston desync dupes
+                blockposition3 = (BlockPosition) list.get(k);
+                iblockdata1 = world.getType(blockposition3);
                 blockposition3 = blockposition3.shift(enumdirection1);
                 map.remove(blockposition3);
                 world.setTypeAndData(blockposition3, (IBlockData) Blocks.MOVING_PISTON.getBlockData().set(BlockPiston.FACING, enumdirection), 68);
-                // Paper start - fix a variety of piston desync dupes
-                if (!allowDesync) {
-                    iblockdata1 = world.getType(oldPos);
-                    map.replace(oldPos, iblockdata1);
-                }
-                world.setTileEntity(blockposition3, BlockPistonMoving.a(allowDesync ? list1.get(k) : iblockdata1, enumdirection, flag, false));
-                if (!allowDesync) {
-                    world.setTypeAndData(oldPos, Blocks.AIR.getBlockData(), 2 | 4 | 16 | 1024); // set air to prevent later physics updates from seeing this block
-                }
-                // Paper end - fix a variety of piston desync dupes
+                world.setTileEntity(blockposition3, BlockPistonMoving.a((IBlockData) list1.get(k), enumdirection, flag, false));
                 aiblockdata[j++] = iblockdata1;
             }
 
diff --git a/src/main/java/net/minecraft/server/TileEntityPiston.java b/src/main/java/net/minecraft/server/TileEntityPiston.java
index b0d52d7405ca24938ca753897f89e6fad575d04d..81c9b5b479f2e62e5f4a6a0212a8f9e9a74725c1 100644
--- a/src/main/java/net/minecraft/server/TileEntityPiston.java
+++ b/src/main/java/net/minecraft/server/TileEntityPiston.java
@@ -258,7 +258,7 @@ public class TileEntityPiston extends TileEntity implements ITickable {
                     IBlockData iblockdata = Block.b(this.a, (GeneratorAccess) this.world, this.position);
 
                     if (iblockdata.isAir()) {
-                    this.world.setTypeAndData(this.position, this.a, com.destroystokyo.paper.PaperConfig.allowPistonDuplication ? 84 : (84 | 2)); // Paper - force notify (flag 2), it's possible the set type by the piston block (which doesn't notify) set this block to air
+                        this.world.setTypeAndData(this.position, this.a, 84);
                         Block.a(this.a, iblockdata, this.world, this.position, 3);
                     } else {
                         if (iblockdata.b(BlockProperties.C) && (Boolean) iblockdata.get(BlockProperties.C)) {
