From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hugo Planque <hookwood01@gmail.com>
Date: Sun, 24 Jan 2021 23:42:56 +0100
Subject: [PATCH] lithium gen


diff --git a/src/main/java/net/minecraft/server/AreaLazy.java b/src/main/java/net/minecraft/server/AreaLazy.java
index 9d9d627973aaa5ace2684524a0fa68137eaea7f0..195276ad4d9a2ae0b73fa2353b4af82835cfd091 100644
--- a/src/main/java/net/minecraft/server/AreaLazy.java
+++ b/src/main/java/net/minecraft/server/AreaLazy.java
@@ -51,6 +51,7 @@ public final class AreaLazy implements Area {
     */
     // Paper end
 
+    public int sample(int i, int j){ return a(i, j); } // Yatopia - OBFHELPER
     @Override
     public int a(int i, int j) {
         long k = ChunkCoordIntPair.pair(i, j);
diff --git a/src/main/java/net/minecraft/server/GenLayer.java b/src/main/java/net/minecraft/server/GenLayer.java
index 805ec502886d31d3d18a48a37e6e4bcf984a8357..6c028bb819c8db46c37afa5fb0fadad670d19c8f 100644
--- a/src/main/java/net/minecraft/server/GenLayer.java
+++ b/src/main/java/net/minecraft/server/GenLayer.java
@@ -7,13 +7,14 @@ public class GenLayer {
 
     private static final Logger LOGGER = LogManager.getLogger();
     private final AreaLazy b;
-
+    private final ThreadLocal<AreaLazy> tlSampler; // Yatopia - Port lithium
     public GenLayer(AreaFactory<AreaLazy> areafactory) {
         this.b = (AreaLazy) areafactory.make();
+        this.tlSampler = ThreadLocal.withInitial(areafactory::make); // Yatopia - Port lithium
     }
 
     public BiomeBase a(IRegistry<BiomeBase> iregistry, int i, int j) {
-        int k = this.b.a(i, j);
+        int k = this.tlSampler.get().sample(i, j); // Yatopia - Port lithium
         ResourceKey<BiomeBase> resourcekey = BiomeRegistry.a(k);
 
         if (resourcekey == null) {
