From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Sat, 20 Feb 2021 22:39:16 +0800
Subject: [PATCH] fixup! Threaded WorldGen


diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 95fc35fec32fed74c93ac1447e6f34e806e0d007..4eb140aad0327165030217399f32e646b4777ab3 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -815,16 +815,18 @@ public class ChunkProviderServer extends IChunkProvider {
 
     // Yatopia start - reduce scheduling overhead
     final ConcurrentLinkedQueue<Runnable> poiQueue = new ConcurrentLinkedQueue<>();
-    private boolean runPoiQueueNext() { // TODO flush poi queue?
-        try {
-            final Runnable runnable = poiQueue.poll();
-            if (runnable != null)
+    private boolean runPoiQueueNext() {
+        boolean hasWork = false;
+        Runnable runnable;
+        while ((runnable = poiQueue.poll()) != null) {
+            hasWork = true;
+            try {
                 runnable.run();
-            return runnable != null;
-        } catch (Throwable t) {
-            t.printStackTrace();
+            } catch (Throwable t) {
+                t.printStackTrace();
+            }
         }
-        return false;
+        return hasWork;
     }
 
     public boolean runTasks() {
