From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Fri, 19 Mar 2021 10:06:16 +0000
Subject: [PATCH] remove pointless bitwise operations


diff --git a/src/main/java/net/minecraft/core/SectionPosition.java b/src/main/java/net/minecraft/core/SectionPosition.java
index 716f91246c4a45fd49af806afd1781f11ff3b346..b5581a39855a23b57c0bcedd5f5213ee892990ef 100644
--- a/src/main/java/net/minecraft/core/SectionPosition.java
+++ b/src/main/java/net/minecraft/core/SectionPosition.java
@@ -68,7 +68,7 @@ public class SectionPosition extends BaseBlockPosition {
     }
 
     public static int b(short short0) {
-        return short0 >>> 0 & 15;
+        return short0 & 15;
     }
 
     public static int c(short short0) {
@@ -97,7 +97,7 @@ public class SectionPosition extends BaseBlockPosition {
 
     public static int unpackX(long i){ return b(i); } // Yatopia - OBFHELPER
     public static int b(long i) {
-        return (int) (i << 0 >> 42);
+        return (int) (i >> 42);
     }
 
     public static int unpackY(long i){ return c(i); } // Yatopia - OBFHELPER
diff --git a/src/main/java/net/minecraft/world/entity/EntityLiving.java b/src/main/java/net/minecraft/world/entity/EntityLiving.java
index dc15593b8222b5e912a757e98fdf162d063e8d44..e7ae01ee38f155ba274578f81c90cb6648a22a81 100644
--- a/src/main/java/net/minecraft/world/entity/EntityLiving.java
+++ b/src/main/java/net/minecraft/world/entity/EntityLiving.java
@@ -898,7 +898,7 @@ public abstract class EntityLiving extends Entity {
             if (flag1 && i > 0) {
                 double d0 = (double) (i >> 16 & 255) / 255.0D;
                 double d1 = (double) (i >> 8 & 255) / 255.0D;
-                double d2 = (double) (i >> 0 & 255) / 255.0D;
+                double d2 = (double) (i & 255) / 255.0D;
 
                 this.world.addParticle(flag ? Particles.AMBIENT_ENTITY_EFFECT : Particles.ENTITY_EFFECT, this.d(0.5D), this.cF(), this.g(0.5D), d0, d1, d2);
             }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/EntityTippedArrow.java b/src/main/java/net/minecraft/world/entity/projectile/EntityTippedArrow.java
index c7c4dd668a273d86d94cc288a0bf8083b8e599b3..8fc76285c8f8d325c9852b5b3103e9d1c4378035 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/EntityTippedArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/EntityTippedArrow.java
@@ -129,7 +129,7 @@ public class EntityTippedArrow extends EntityArrow {
         if (j != -1 && i > 0) {
             double d0 = (double) (j >> 16 & 255) / 255.0D;
             double d1 = (double) (j >> 8 & 255) / 255.0D;
-            double d2 = (double) (j >> 0 & 255) / 255.0D;
+            double d2 = (double) (j & 255) / 255.0D;
 
             for (int k = 0; k < i; ++k) {
                 this.world.addParticle(Particles.ENTITY_EFFECT, this.d(0.5D), this.cF(), this.g(0.5D), d0, d1, d2);
diff --git a/src/main/java/net/minecraft/world/item/alchemy/PotionUtil.java b/src/main/java/net/minecraft/world/item/alchemy/PotionUtil.java
index 795bc60a73e1e628590803fd515ffb78302d4f97..dcf025753206c9fa233ed9ad2881bdfbaa7141e4 100644
--- a/src/main/java/net/minecraft/world/item/alchemy/PotionUtil.java
+++ b/src/main/java/net/minecraft/world/item/alchemy/PotionUtil.java
@@ -97,7 +97,7 @@ public class PotionUtil {
 
                     f += (float) (l * (k >> 16 & 255)) / 255.0F;
                     f1 += (float) (l * (k >> 8 & 255)) / 255.0F;
-                    f2 += (float) (l * (k >> 0 & 255)) / 255.0F;
+                    f2 += (float) (l * (k & 255)) / 255.0F;
                     j += l;
                 }
             }
