From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Tue, 2 Mar 2021 20:39:30 -0500
Subject: [PATCH] print Clientbrand on login


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index bea7801f795268922e9dbf413e005c1bf1d2b882..c407ba3e783d9d9da895d3f334fbe73bae343d35 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -3050,6 +3050,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
             }
         }
 
+        System.out.println("Client brand: "+this.clientBrandName);
     }
 
     // Paper start - brand support
