From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Simon Gardling <titaniumtown@gmail.com>
Date: Tue, 2 Mar 2021 20:39:30 -0500
Subject: [PATCH] print Clientbrand on login


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 9016554f9f7b637ce5e80ce56504b4a23b25e7ac..8b4d76da5dd2503cd149e2171d8a2efdc5ac429c 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -3056,6 +3056,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
             }
         }
 
+        System.out.println("Client brand: "+this.clientBrandName);
     }
 
     // Paper start - brand support
